### PR TITLE
Refactor doc.populate FN type.

### PR DESCRIPTION
**Pull request by mohammad0-0ahmad**
Related to : https://github.com/Automattic/mongoose/issues/11598

### CODE REVIEW

File: types/document.d.ts [hunk 1]
```diff
@@ -186,10 +186,8 @@ declare module 'mongoose' {
     $parent(): Document | undefined;
 
     /** Populates document references. */
-    populate<Paths = {}>(path: string | PopulateOptions | (string | PopulateOptions)[], callback: Callback<this & Paths>): void;
-    populate<Paths = {}>(path: string, names: string, callback: Callback<this & Paths>): void;
-    populate<Paths = {}>(path: string, names: string): Promise<this & Paths>;
-    populate<Paths = {}>(path: string | PopulateOptions | (string | PopulateOptions)[]): Promise<this & Paths>;
+    populate<Paths = {}>(path: string | PopulateOptions | (string | PopulateOptions)[], select?: string | AnyObject, model?: Model<unknown>, match?: AnyObject, options?: PopulateOptions, callback?: Callback<this & Paths>): void;
```
Comment (mohammad0-0ahmad) [inline types/document.d.ts:189]: Should option argument be existing?

Comment (Uzlopak) [reply types/document.d.ts:189]: @vkarpov15

File: test/types/document.test.ts [hunk 2]
```diff
@@ -164,4 +175,9 @@ function gh11435() {
   ItemSchema.pre('validate', function preValidate() {
     expectType<Model<unknown>>(this.model('Item1'));
   });
+}
+
+async function gh11598() {
+  const doc = await Test.findOne();
```
Comment (vkarpov15) [inline test/types/document.test.ts:181]: You can do `await Test.findOne().orFail()` and avoid the need for `?.` below

File: types/document.d.ts (hunk 1 above)
Comment (vkarpov15) [reply types/document.d.ts:189]: Yes it should.

File: types/document.d.ts [hunk 3]
```diff
@@ -186,10 +186,8 @@ declare module 'mongoose' {
     $parent(): Document | undefined;
 
     /** Populates document references. */
-    populate<Paths = {}>(path: string | PopulateOptions | (string | PopulateOptions)[], callback: Callback<this & Paths>): void;
-    populate<Paths = {}>(path: string, names: string, callback: Callback<this & Paths>): void;
```
Comment (vkarpov15) [inline types/document.d.ts:190]: Don't get rid of these. The issue is that, if there's more than 1 non-callback arg, `path` must be a string. We don't support `populate(PopulateOptions, string, Model<unknown>)`, for example. So keep this syntax.

**Review by vkarpov15 — CHANGES_REQUESTED** [review]

**Comment by mohammad0-0ahmad** [comment]
@vkarpov15 
I've noticed that when rebasing this PR on top of master branch, the model loose and doc for example lose its type & become any.
Although, all tests are passed.
Can u please double-check this?
Untitled

**Review by vkarpov15 — APPROVED** [review]
Nice work, thanks :+1:

**Pull request merged**